ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java

```diff
@@ -0,0 +1,535 @@
+      }
+    }
+
+    void initializeTopics(final KsqlEngine ksqlEngine) {
+      for (Topic topic : topics) {
+        ksqlEngine.getTopicClient().createTopic(topic.getName(), 1, (short) 1);
+        if (topic.getSchema() != null) {
+          try {
+            ksqlEngine.getSchemaRegistryClient().register(
+                topic.getName() + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX, topic.getSchema());
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    }
+  }
+
+  private static TopologyTestDriver buildStreamsTopology(final Query query,
+                                                         final KsqlEngine ksqlEngine,
+                                                         final Properties streamsProperties) {
+    final List<QueryMetadata> queries = new ArrayList<>();
+    query.statements().forEach(
+        q -> queries.addAll(ksqlEngine.buildMultipleQueries(q, Collections.emptyMap()))
+    );
+    return new TopologyTestDriver(queries.get(queries.size() - 1).getTopology(),
+        streamsProperties,
+        0);
+  }
+
+  static List<String> findTests(final String dir) throws IOException {
+    final List<String> tests = new ArrayList<>();
+    try (final BufferedReader reader =
+             new BufferedReader(
+                 new InputStreamReader(EndToEndEngineTestUtil.class.getClassLoader().
+                     getResourceAsStream(dir)))) {
+
+      String test;
+      while ((test = reader.readLine()) != null) {
+        if (test.endsWith(".json")) {
+          tests.add(test);
+        }
+      }
+    }
+    return tests;
+  }
+
+  static void shouldBuildAndExecuteQuery(final Query query) {
+    final MetaStore metaStore = new MetaStoreImpl(new InternalFunctionRegistry());
+    final Map<String, Object> config = new HashMap<String, Object>() {{
+      put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
+      put("application.id", "KSQL-TEST");
+      put("commit.interval.ms", 0);
+      put("cache.max.bytes.buffering", 0);
+      put("auto.offset.reset", "earliest");
+      put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+    }};
+    final Properties streamsProperties = new Properties();
+    final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
+
+    streamsProperties.putAll(config);
+
+    try (final KsqlEngine ksqlEngine = new KsqlEngine(
+        new KsqlConfig(config),
+        new FakeKafkaTopicClient(),
+        schemaRegistryClient,
+        metaStore
+    )) {
+      query.initializeTopics(ksqlEngine);
+      final TopologyTestDriver testDriver = buildStreamsTopology(query, ksqlEngine, streamsProperties);
+      query.processInput(testDriver, schemaRegistryClient);
+      query.verifyOutput(testDriver, schemaRegistryClient);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  static Object valueSpecToAvro(final Object spec, final org.apache.avro.Schema schema) {
```

Comment:
For this and the next method, the input type is `Object` and the output is `Object`. What actual transformation is happening here? A javadoc comment would help. For instance, is the value spec type some Connect data type?

Reply:
I'll add a comment. Basically its taking the spec for the contents of a record in the form of maps, lists, and primitives and converting it to an avro object.